File: honda-hr-frontend/src/firebase.ts
/**
 * Sign-in, through Firebase. Two routes: Google, and email + password.
 *
 * Firebase holds the passwords and sends the emails — the confirmation link,
 * the reset link. Our own server holds none of that. The reason is not tidiness
 * but delivery: sending mail ourselves needs a provider that will write to an
 * arbitrary recipient, and the free tier of the one configured refuses every
 * address except the account holder's until a domain is verified, which needs a
 * domain nobody here owns. Firebase sends from its own infrastructure, to
 * anyone, at no cost.
 *
 * What Firebase decides is one question only: does this person control this
 * mailbox. Whether that mailbox may see a candidate's CV is decided by our own
 * server, against the allowed-users list. A verified address has never been
 * permission here and must not become it.
 *
 * The SDK is loaded from Google's CDN at the moment it is first needed, rather
 * than installed as an npm dependency. Two reasons, in order of importance:
 *
 *   1. Nothing is lost by it. Firebase sign-in already cannot work offline —
 *      the browser has to talk to Google's identity servers either way.
 *      Fetching the SDK from the same origin adds no dependency that was not
 *      already there.
 *   2. The npm package is `firebase`, an umbrella over a dozen products we do
 *      not use, and its deeply nested paths were failing to install on
 *      Windows. Not shipping it also keeps the build reproducible on a machine
 *      where that install refuses to complete.
 *
 * The trade is that a typo in an export name becomes a runtime error rather
 * than a compile error, since a URL import carries no types. The surface used
 * here is a handful of functions wide and pinned to one SDK version, which is a
 * small enough target to be worth it.
 *
 * The config values below are not secrets. A Firebase web config ships inside
 * every page that uses it and is visible to anyone; what protects the project
 * is the authorised-domains list in the console, plus the fact that our server
 * checks the list. The service-account key, which IS a secret, lives only on
 * the server and never reaches this file.
 */

/** Pinned deliberately: "latest" would let Google change our app underneath us. */
const SDK_VERSION = "12.18.0";
const CDN = `https://www.gstatic.com/firebasejs/${SDK_VERSION}`;

const config = {
  apiKey: import.meta.env.VITE_FIREBASE_API_KEY,
  authDomain: import.meta.env.VITE_FIREBASE_AUTH_DOMAIN,
  projectId: import.meta.env.VITE_FIREBASE_PROJECT_ID,
  appId: import.meta.env.VITE_FIREBASE_APP_ID,
};

/** Whether this build has enough configuration to attempt Firebase sign-in. */
export const firebaseConfigured = (): boolean =>
  Boolean(config.apiKey && config.authDomain && config.projectId && config.appId);

/* eslint-disable @typescript-eslint/no-explicit-any */
type Sdk = { app: any; auth: any };

let loading: Promise<Sdk> | null = null;

/**
 * Fetch the two SDK modules, once per page load.
 *
 * A failed load is deliberately not cached: the usual cause is a dropped
 * connection, and a second attempt a moment later should be allowed to
 * succeed rather than replaying the first failure forever.
 */
function loadSdk(): Promise<Sdk> {
  if (!loading) {
    loading = Promise.all([
      import(/* @vite-ignore */ `${CDN}/firebase-app.js`),
      import(/* @vite-ignore */ `${CDN}/firebase-auth.js`),
    ])
      .then(([app, auth]) => ({ app, auth }))
      .catch((e) => {
        loading = null;
        throw new SdkUnavailable(String(e));
      });
  }
  return loading;
}

/** The SDK itself could not be fetched — distinct from Firebase refusing us. */
class SdkUnavailable extends Error {
  readonly code = "sdk/unavailable";
}

/* ---------------------------------------------------------------------------
 * Google sign-in
 *
 * The shorter of the two routes: no password to remember, no link to find.
 * Most of the people on the list already hold the Google account for the
 * address they were added under, so one click is the whole flow.
 *
 * It is not a second front door. Google proves the mailbox — exactly what the
 * confirmation email proves — and the server then puts that address to the
 * same allowed-users list. Somebody with a Google account and no place on that
 * list gets the same refusal they would get anywhere else.
 * ------------------------------------------------------------------------ */

/** Sign in with Google and return the Firebase ID token for our server. */
export async function signInWithGoogle(): Promise<string> {
  const sdk = await loadSdk();
  const { initializeApp, getApps } = sdk.app;
  const { getAuth, GoogleAuthProvider, signInWithPopup, signOut } = sdk.auth;

  const app = getApps().length ? getApps()[0] : initializeApp(config);
  const auth = getAuth(app);
  auth.useDeviceLanguage();

  const provider = new GoogleAuthProvider();
  // Always ask which account. Without this, somebody already signed into a
  // personal Google account in this browser is silently signed in as that
  // one — and then refused, with no clue that the wrong account was used.
  provider.setCustomParameters({ prompt: "select_account" });

  const credential = await signInWithPopup(auth, provider);
  const token = await credential.user.getIdToken();
  // We have what we need. Firebase's own session is not ours.
  await signOut(auth).catch(() => undefined);
  return token;
}

/** Turn a Google sign-in failure into something worth showing a person. */
export function googleErrorMessage(err: unknown): string {
  const code = (err as { code?: string })?.code ?? "";
  switch (code) {
    case "sdk/unavailable":
      return "Could not reach Google to start sign-in. Check the connection and try again.";
    case "auth/popup-blocked":
      return "Your browser blocked the Google window. Allow pop-ups for this site, then try again.";
    case "auth/popup-closed-by-user":
    case "auth/cancelled-popup-request":
      // Not an error worth alarming anybody about: they closed it themselves.
      return "";
    case "auth/operation-not-allowed":
      return "Google sign-in is not switched on in Firebase. In the console: Authentication → Sign-in method → add Google.";
    case "auth/unauthorized-domain":
      return "This address is not on the Firebase authorised-domains list. An administrator has to add it in the Firebase console.";
    default:
      return err instanceof Error ? err.message : "Could not sign in with Google.";
  }
}

/* ---------------------------------------------------------------------------
 * Email and password, held by Firebase
 *
 * The password lives in Firebase, not in our database. That is the whole point
 * of routing it here: one place decides whether a person is who they say they
 * are, and our server only ever asks "is this proven address on the list?".
 *
 * `email_verified` is not optional on this route, and the reason is worth
 * stating plainly. Firebase lets anybody register ANY address — proving you
 * own it is a separate step. So a stranger could sign up as a colleague's
 * work address, and the allowlist would happily wave them through, because
 * the address really is on it. The verification email is what closes that.
 * ------------------------------------------------------------------------ */

/** Where Firebase sends people back to after they confirm the address. */
function returnUrl(): string {
  return `${window.location.origin}${window.location.pathname}#/signin`;
}

/**
 * Create the Firebase account and send the confirmation email.
 *
 * Throws `auth/email-already-in-use` when the account exists — the caller
 * treats that as "you already signed up", not as a failure.
 */
export async function signUpWithPassword(address: string, password: string): Promise<void> {
  const sdk = await loadSdk();
  const { initializeApp, getApps } = sdk.app;
  const { getAuth, createUserWithEmailAndPassword, sendEmailVerification, signOut } = sdk.auth;

  const app = getApps().length ? getApps()[0] : initializeApp(config);
  const auth = getAuth(app);
  auth.useDeviceLanguage();

  const credential = await createUserWithEmailAndPassword(auth, address, password);
  await sendEmailVerification(credential.user, { url: returnUrl() });
  // They are not signed in until they confirm, so do not leave a half-session
  // sitting in the browser looking like success.
  await signOut(auth).catch(() => undefined);
}

/** Ask Firebase to send the confirmation email again. */
export async function resendVerification(address: string, password: string): Promise<void> {
  const sdk = await loadSdk();
  const { initializeApp, getApps } = sdk.app;
  const { getAuth, signInWithEmailAndPassword, sendEmailVerification, signOut } = sdk.auth;

  const app = getApps().length ? getApps()[0] : initializeApp(config);
  const auth = getAuth(app);
  const credential = await signInWithEmailAndPassword(auth, address, password);
  await sendEmailVerification(credential.user, { url: returnUrl() });
  await signOut(auth).catch(() => undefined);
}

/** Thrown when the password is right but the address was never confirmed. */
export class EmailNotVerified extends Error {
  readonly code = "auth/email-not-verified";
}

/** Sign in, and return the Firebase ID token for our server. */
export async function signInWithPassword(address: string, password: string): Promise<string> {
  const sdk = await loadSdk();
  const { initializeApp, getApps } = sdk.app;
  const { getAuth, signInWithEmailAndPassword, sendEmailVerification, signOut } = sdk.auth;

  const app = getApps().length ? getApps()[0] : initializeApp(config);
  const auth = getAuth(app);
  auth.useDeviceLanguage();

  const credential = await signInWithEmailAndPassword(auth, address, password);
  if (!credential.user.emailVerified) {
    // Send another one on the way out: somebody who reaches this point has
    // lost or never received the first, and telling them to look for an email
    // that is not there is the least useful thing this could do.
    await sendEmailVerification(credential.user, { url: returnUrl() }).catch(() => undefined);
    await signOut(auth).catch(() => undefined);
    throw new EmailNotVerified(
      "Confirm your email address first. We have sent the link again — check your inbox, and the spam folder.",
    );
  }

  const token = await credential.user.getIdToken();
  await signOut(auth).catch(() => undefined);
  return token;
}

/** Firebase emails the reset link itself — free, to any address. */
export async function sendPasswordReset(address: string): Promise<void> {
  const sdk = await loadSdk();
  const { initializeApp, getApps } = sdk.app;
  const { getAuth, sendPasswordResetEmail } = sdk.auth;

  const app = getApps().length ? getApps()[0] : initializeApp(config);
  const auth = getAuth(app);
  auth.useDeviceLanguage();
  await sendPasswordResetEmail(auth, address, { url: returnUrl() });
}

/** Turn a Firebase password error into something worth showing a person. */
export function passwordErrorMessage(err: unknown): string {
  const code = (err as { code?: string })?.code ?? "";
  switch (code) {
    case "sdk/unavailable":
      return "Could not reach Google to sign in. Check the connection and try again.";
    case "auth/email-not-verified":
      return err instanceof Error ? err.message : "Confirm your email address first.";
    case "auth/invalid-credential":
    case "auth/wrong-password":
    case "auth/user-not-found":
      // One message for both, deliberately: separate ones tell a stranger
      // which addresses have accounts here.
      return "That email address and password do not match.";
    case "auth/email-already-in-use":
      return "That address already has a password. Use Sign in, or Forgot password.";
    case "auth/weak-password":
      return "That password is too weak. Use at least 10 characters.";
    case "auth/invalid-email":
      return "That does not look like an email address.";
    case "auth/too-many-requests":
      return "Too many attempts from this device. Wait a few minutes, then try again.";
    case "auth/operation-not-allowed":
      return "Email and password sign-in is not switched on in Firebase. In the console: Authentication → Sign-in method → Email/Password.";
    case "auth/unauthorized-domain":
      return "This address is not on the Firebase authorised-domains list. An administrator has to add it.";
    default:
      return err instanceof Error ? err.message : "Could not sign in.";
  }
}

File: cv-model/api/firebase.py
"""
Phone sign-in, verified through Firebase.

Firebase sends the SMS and checks the code, which is worth having: sending SMS
to Pakistani networks otherwise means a gateway account, a registered sender ID
and a business relationship nobody has yet.

What Firebase does NOT do is decide who may use this system. It answers one
question — does this phone number belong to the person holding the browser —
and the answer is a signed token. Whether that number is allowed anywhere near
a candidate's CV is decided in api/auth.py, against the same list the email
route uses. Treating a valid Firebase token as permission would admit anybody
in the world who can receive an SMS.
"""

from __future__ import annotations

import json
import logging
import threading
from pathlib import Path
from typing import Any

from .config import auth_settings as cfg

log = logging.getLogger("api.firebase")

_app: Any = None
_lock = threading.Lock()


class FirebaseUnavailable(RuntimeError):
    """Phone sign-in is not configured, or the library is not installed."""


class BadPhoneToken(RuntimeError):
    """The token was not accepted, or carries no verified phone number."""


class BadEmailToken(RuntimeError):
    """The token was not accepted, or carries no verified email address."""


def _credentials() -> Any:
    """Read the service account, from a file path or from the JSON itself.

    A deployment usually cannot mount a file, so the whole JSON arrives in an
    environment variable; a laptop usually has the downloaded file. Supporting
    both means the same setting works in both places.
    """
    from firebase_admin import credentials

    raw = cfg.firebase_credentials.strip()
    if not raw:
        raise FirebaseUnavailable("FIREBASE_CREDENTIALS is not set.")

    if raw.startswith("{"):
        try:
            return credentials.Certificate(json.loads(raw))
        except json.JSONDecodeError as e:
            raise FirebaseUnavailable(
                f"FIREBASE_CREDENTIALS looks like JSON but could not be parsed: {e}"
            ) from e

    path = Path(raw)
    if not path.is_file():
        raise FirebaseUnavailable(
            f"FIREBASE_CREDENTIALS points at {raw!r}, which is not a file. "
            "Give the path to the service account JSON, or paste the JSON itself."
        )
    return credentials.Certificate(str(path))


def _get_app() -> Any:
    global _app
    if _app is not None:
        return _app
    with _lock:
        if _app is not None:
            return _app
        try:
            import firebase_admin
        except ImportError as e:
            raise FirebaseUnavailable(
                "Phone sign-in needs the firebase-admin package: pip install firebase-admin"
            ) from e

        _app = firebase_admin.initialize_app(_credentials(), name="cv-shortlist")
        log.info("Firebase phone sign-in is active.")
        return _app


def verified_phone_number(id_token: str) -> str:
    """Return the phone number Firebase confirmed, or raise.

    The token is checked against Google's signing keys — it cannot be forged by
    the browser, which is the whole reason the check happens here rather than
    trusting a number the page sends us.
    """
    from firebase_admin import auth as fb_auth

    app = _get_app()
    try:
        claims = fb_auth.verify_id_token(id_token, app=app)
    except Exception as e:  # noqa: BLE001 — the SDK raises several types
        raise BadPhoneToken(f"That sign-in could not be verified: {e}") from e

    number = (claims.get("phone_number") or "").strip()
    if not number:
        # An email or anonymous Firebase login would also produce a valid
        # token. Accepting one here would let somebody sign in with a Google
        # account and be treated as a verified phone.
        raise BadPhoneToken("That sign-in did not include a verified phone number.")
    return number


def verified_email(id_token: str) -> tuple[str, str]:
    """Return (email address, which provider proved it), or raise.

    This exists because our own mail route cannot reach anybody. Sending a
    one-time code ourselves needs a mail provider that will deliver to an
    arbitrary recipient, and the free tier of the one configured refuses every
    address except the account holder's until a domain is verified — which
    needs a domain nobody here owns. Firebase sends the sign-in link from its
    own infrastructure, to anyone, at no cost.

    `email_verified` is checked, not just `email`. A Firebase project can also
    hold email/password accounts whose address was never confirmed; accepting
    one would let somebody register any address they liked in the Firebase
    console and walk in as that person. The claim is only true after Firebase
    itself has proved the address, which is exactly what the link does.

    The provider is returned as well as the address because the audit trail
    should say HOW somebody got in, not merely that they did. A Google sign-in
    and a one-time emailed link are different events with different failure
    modes, and a log that flattens them into "signed in" cannot answer the
    question it exists for.

    Any Firebase provider that proves an address is accepted — a link, Google,
    whatever is enabled in the console. That is deliberate: they all answer the
    same one question, and the answer to a DIFFERENT question, may this person
    see a CV, is still decided in auth.py against the allowed-users list. A
    verified mailbox has never been permission here and must not become it.
    """
    from firebase_admin import auth as fb_auth

    app = _get_app()
    try:
        claims = fb_auth.verify_id_token(id_token, app=app)
    except Exception as e:  # noqa: BLE001 — the SDK raises several types
        raise BadEmailToken(f"That sign-in could not be verified: {e}") from e

    address = (claims.get("email") or "").strip()
    if not address:
        raise BadEmailToken("That sign-in did not include an email address.")
    if not claims.get("email_verified"):
        raise BadEmailToken("That email address has not been confirmed.")

    provider = str((claims.get("firebase") or {}).get("sign_in_provider") or "firebase")
    return address, provider
